sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/AADClientRegistrationRepository.java

```diff
@@ -3,83 +3,402 @@
+                + ".authorization-grant-type must be configured to 'authorization_code'.");
+        }
+    }
+
+    private ClientRegistration.Builder toClientRegistrationBuilder(String registrationId,
+                                                                   AuthorizationClientProperties clientProperties) {
+        // todo check null
+        Assert.notNull(clientProperties.getAuthorizationGrantType(),
+            "AuthorizationGrantType can not be empty. " + "registrationId: " + registrationId + ".");
+        LOGGER.debug("Client {} AuthorizationClientProperties: {}.", registrationId, clientProperties);
+        AADAuthorizationServerEndpoints endpoints =
+            new AADAuthorizationServerEndpoints(properties.getBaseUri(), properties.getTenantId());
+        ClientRegistration.Builder clientRegistrationBuilder;
+        switch (clientProperties.getAuthorizationGrantType()) {
+            case AUTHORIZATION_CODE:
+                clientRegistrationBuilder = toAuthorizationCodeBuilder(registrationId, endpoints, clientProperties);
+                break;
+            case ON_BEHALF_OF:
+                clientRegistrationBuilder = toOnBehalfOfBuilder(registrationId, endpoints, clientProperties);
+                break;
+            case CLIENT_CREDENTIALS:
+                clientRegistrationBuilder = toClientCredentialsBuilder(registrationId, endpoints, clientProperties);
+                break;
+            default:
+                throw new IllegalArgumentException("Unsupported authorization type "
+                    + clientProperties.getAuthorizationGrantType().getValue());
+        }
+        return clientRegistrationBuilder;
+    }
+
+    private ClientRegistration.Builder toAuthorizationCodeBuilder(String registrationId,
+                                                              AADAuthorizationServerEndpoints endpoints,
+                                                              AuthorizationClientProperties clientProperties) {
+        return ClientRegistration.withRegistrationId(registrationId)
+                                 .clientName(registrationId)
+                                 .authorizationGrantType(AuthorizationGrantType.AUTHORIZATION_CODE)
+                                 .scope(toScopes(clientProperties))
+                                 .redirectUri(properties.getRedirectUriTemplate())
+                                 .userNameAttributeName(properties.getUserNameAttribute())
+                                 .clientId(toClientId(properties))
+                                 .clientSecret(properties.getClientSecret())
+                                 .authorizationUri(endpoints.authorizationEndpoint())
+                                 .tokenUri(endpoints.tokenEndpoint())
+                                 .jwkSetUri(endpoints.jwkSetEndpoint())
+                                 .providerConfigurationMetadata(providerConfigurationMetadata(endpoints));
```

Comment:
Can we reuse this part of code, for example:

```
 private ClientRegistration.Builder toClientRegistration(String registrationId,
                                                              AuthorizationGrantType type,
                                                              AADAuthorizationServerEndpoints endpoints,
                                                              AuthorizationClientProperties clientProperties) {
```

Reply:
Yes, but there will still be the branch judgement because there are different types of property entries that need to be set